appcues/src/main/java/com/appcues/ui/ExperienceStepFormState.kt

```diff
@@ -1,39 +1,102 @@
 package com.appcues.ui
 
 import androidx.compose.runtime.mutableStateOf
+import com.appcues.data.model.ExperiencePrimitive.OptionSelectPrimitive
+import com.appcues.data.model.ExperiencePrimitive.TextInputPrimitive
+import com.appcues.ui.ExperienceStepFormItemState.OptionSelectFormItemState
+import com.appcues.ui.ExperienceStepFormItemState.TextInputFormItemState
 import java.util.UUID
 
 internal class ExperienceStepFormState {
-    private var formItems: HashMap<UUID, ExperienceStepFormItem> = hashMapOf()
+    private var _formItems: HashMap<UUID, ExperienceStepFormItemState> = hashMapOf()
+
+    private var itemIndex = 0
+
+    val formItems: Collection<ExperienceStepFormItemState>
+        get() = _formItems.values.sortedBy { it.index }
 
     // this can be used by buttons needing to know if they are enabled or not based on required input
     val isFormComplete = mutableStateOf(true)
 
-    fun captureFormItem(id: UUID, item: ExperienceStepFormItem) {
-        formItems[id] = item
-        isFormComplete.value = !formItems.values.any { !it.isComplete }
+    fun getValue(primitive: TextInputPrimitive) =
+        getItem(primitive).value
+
+    fun setValue(primitive: TextInputPrimitive, value: String) {
+        val item = getItem(primitive)
+        item.value = value
+        updateFormItem(item)
+    }
+
+    fun getValue(primitive: OptionSelectPrimitive) =
+        getItem(primitive).values
+
+    fun setValue(primitive: OptionSelectPrimitive, values: Set<String>) {
+        val item = getItem(primitive)
+        item.values = values
+        updateFormItem(item)
+    }
+
+    private fun getItem(primitive: TextInputPrimitive): TextInputFormItemState {
+        var item = _formItems[primitive.id] as? TextInputFormItemState
+        if (item != null) return item
+
+        // create new state tracking object
+        item = with(primitive) {
+            TextInputFormItemState(itemIndex++, id, "textInput", label.text, required, defaultValue ?: "")
```

Comment:
this `itemIndex++` is an attempt to keep an ordered set of form inputs so the analytics can be ordered in the same fashion as the JSON structure - it _seems_ to work, but I'm not sure if always guaranteed to do so.

Reply:
I think it should work, compositions happen in the same thread and its linear.